### PR TITLE
feat: plugin conventions

### PR DESCRIPTION
**Pull request by 3cp**
# Pull Request

## 📖 Description

This is ongoing work to implement Aurelia naming conventions at bundler level for Aurelia 2.
Aurelia v1 deals with conventions at runtime, while it's very flexible, it added up tooling complexity and runtime complexity.
Aurelia 2 runtime is very simple. We hook conventions (inject needed code) at bundler level even before the transpiler see the source code.

### 🎫 Issues

#116 Preprocess html view template
Part of #509 naming conventions.

## 👩‍💻 Reviewer Notes

This is an ongoing effort.
The first project is `@aurelia/plugin-conventions`, it's the base to be wrapped by all plugins (webpack, parcel, gulp...).
The plugin-conventions provides pure functions to preprocess js/ts resources and html view template. No fs API used. So it can run inside browser (if we ever need it in an online IDE).

Recommend to review the test code inside `packages/__test__/plugin-conventions`, to see if I covered enough edge cases.

Also, I am not a TS guy, the code didn't use much typing. But we want to get this working first, before improving the implementation.

### Note on html template

Since we preprocess it, we now can tolerate bad html from user.

```html
<import from="a"></import> <!-- this is correct html5 -->
<!-- following are incorrect html5, but we can accept them -->
<import from="a">
<import from="a" />
```

Both `<import>` and `<require>` tags are acceptable, both before or after `<template>`. In fact, we accept them everywhere in the html template, even after the last `</template>`.

Aurelia 2 finally avoided `<require>` + `<td>` issue, because `<require>`(or `<import>`) is not present in final html string, the deps were picked out by preprocessor.

## 📑 Test Plan

The html template preprocessor has option to write some typing for TS code, in order to avoid complain from following transpiling.

The js/ts preprocessor has no option for adding typing. We will see what's needed.

We need full app build to find out possible missed parts (typing is my biggest concern).

## ⏭ Next Steps

- [x] experimental webpack plugin

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
Very excited to see this getting started! 🎉

**Comment by 3cp** [comment]
@EisenbergEffect have a look of the test cases.

File: packages/__tests__/plugin-conventions/name-convention.spec.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  it('gets custom attribute like resource', function() {
+    const result = nameConvention('FooBarCustomAttribute');
+    assert.deepEqual(result, {name: 'foo-bar', type: 'customAttribute'});
+  });
+
+  it('gets value converter like resource', function() {
+    const result = nameConvention('FooBarValueConverter');
+    assert.deepEqual(result, {name: 'foo-bar', type: 'valueConverter'});
```
Comment (EisenbergEffect) [inline packages/__tests__/plugin-conventions/name-convention.spec.ts]: This should actually be `fooBar` due to the fact that we can use a `-` a binding expression without it being interpreted as a minus. Same for the binding behavior convention below (but not binding command).

File: packages/__tests__/plugin-conventions/preprocess-resource.spec.ts [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+export class FooBarCustomAttribute {}
+`;
+    const result = preprocessResource('bar/FooBar.js', code, true);
+    assert.equal(result.code, expected);
+  });
+
+  it('injects valueConverter decorator', function () {
+    const code = `export class FooBarValueConverter {}\n`;
+    const expected = `import { valueConverter } from '@aurelia/runtime';
+@valueConverter('foo-bar')
```
Comment (EisenbergEffect) [inline packages/__tests__/plugin-conventions/preprocess-resource.spec.ts]: See note above about value converter and binding behavior naming.

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 1 above)
Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: Good catch! My mistakes.

**Comment by EisenbergEffect** [comment]
@3cp I left a couple comments about name conventions. Otherwise, looks pretty good :)

File: packages/__tests__/plugin-conventions/name-convention.spec.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { nameConvention } from '@aurelia/plugin-conventions';
+import { assert } from '@aurelia/testing';
+
+describe('nameConvention', function () {
+  it('complains about empty input', function() {
+    assert.throws(() => nameConvention(''));
```
Comment (bigopon) [inline packages/__tests__/plugin-conventions/name-convention.spec.ts]: maybe throw on anything that is potentially invalid? Just to keep the spirit of our testing, started by @fkleuver

File: packages/__tests__/plugin-conventions/name-convention.spec.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  it('gets custom element like resource', function() {
```
Comment (bigopon) [inline packages/__tests__/plugin-conventions/name-convention.spec.ts]: would love to see some number tests as well, sometime i use them 😬

File: packages/__tests__/plugin-conventions/preprocess-html-template.spec.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import { preprocessHtmlTemplate } from '@aurelia/plugin-conventions';
```
Comment (bigopon) [inline packages/__tests__/plugin-conventions/preprocess-html-template.spec.ts]: ❤️ this file

File: packages/__tests__/plugin-conventions/strip-html-import.spec.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+</template>
+`;
+
+    assert.deepEqual(stripHtmlImport(html), {
+      html: expected,
+      deps: ['./a']
+    });
+  });
+
+  it('strips mixed import/require tags with wrong or missing close tag', function() {
```
Comment (bigopon) [inline packages/__tests__/plugin-conventions/strip-html-import.spec.ts]: woohoo welcome to html town 😆

File: packages/plugin-conventions/src/strip-html-import.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { parseFragment, DefaultTreeElement, ElementLocation } from 'parse5';
```
Comment (bigopon) [inline packages/plugin-conventions/src/strip-html-import.ts]: This file looks simple but exciting 🎉

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 3 above)
Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: It's kind of internal API to translate ClassName. BTW, what kind of invalid input you has in mind?

Even `WhatEverName` will be a custom element based on convention.

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 4 above)
Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: What's number tests?

Comment (bigopon) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: Example would be
MyEl1 -> my-el1 or myel-1?
Just need to have a few tests for that i think.

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 3 above)
Comment (bigopon) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: Undefined, false symbol null number

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 4 above)
Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: I see, will add it.
The kebabCase and camelCase is from `@aurelia/kernel`, that's why I didn't bother to test many edge cases.

File: packages/__tests__/plugin-conventions/name-convention.spec.ts (hunk 3 above)
Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: I thought because this is TS project, you don't need to do that kind of tests anymore?

Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: I think the compiler will reject that kind of test code to build.

Comment (3cp) [reply packages/__tests__/plugin-conventions/name-convention.spec.ts]: @fkleuver I tested `nameConvention(null)`, the `npm run dev` didn't complain about the usage against the typing `nameConvention(className: string)` in the test code. Is there any missed typing setup in our test code compilation?

**Comment by EisenbergEffect** [comment]
@3cp Note that I think we’ll want some APIs or method of configuration in the future that let’s people customize the conventions. We don’t need that for the first pass though. I’d love to get everything working and then come back to that in the 2nd or 3rd pass, once we feel confident about the general approach.

**Comment by 3cp** [comment]
I see, that probably will be options passed to AOT first, then AOT might call conventions internally. We will see how it goes.

The current implementation is tiny, easy to change in future.

**Comment by 3cp** [comment]
plugin-webpack is renamed to webpack-loader, because it's not a plugin, it's a loader just like ts-loader. I tested it in babel and ts webpack apps.

@EisenbergEffect @fkleuver let's merge this PR and publish these two packages, so we can test in webpack app more easily.

**Comment by fkleuver** [comment]
@3cp The builds do need to pass first. It's complaining somehow about not being able to find `@babel/parser`? https://circleci.com/gh/aurelia/aurelia/21357?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by 3cp** [comment]
I see, modify-code use babel parse to tokenize all latest js/ts. I will check why it complains.

**Comment by 3cp** [comment]
@fkleuver it looks like those browser test I never did locally. Questions:

1. I guess I need to add some deps to `packages/__test__/package.json`?
2. how do I add plugin-conventions to browser test, but skip webpack-loader on browser?

**Comment by 3cp** [comment]
I don't quite understand why the karma+webpack setup didn't trace and bundle the necessary packages. If this is the first time an `@aurelia` package requires 3rd party npm package, we might have a hole in karma+webpack config.
